### PR TITLE
Theo Initial PR

### PR DESCRIPTION
**Pull request by Choklatier**

### CODE REVIEW

File: backend/main.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+
+manager = ConnectionManager()
+
+game_handler = GameHandler()
+game = game_handler.get_game()
+
+# Server loop
+def loop():
+    while True:
```
Comment (frankbryden) [inline backend/main.py:45]: This is gonna sit in a very tight loop doing nothing most of the time. We need to find a slightly better way of doing this.
In the case when no one is in the lobby (majority of time), we need some way of blocking the thread until a player joins.

In the case when a game is being played, we can probably get a way with doing 10 or 100 iterations per second.

Comment (frankbryden) [reply backend/main.py:45]: For now perhaps a sleep will suffice

File: backend/main.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        current_players = game_handler.players
+        
+        if len(current_players) > 0 and not game.ongoing:
+            game.start_game()
+            print("Game starting!")
+        
+        if game.is_crashed():
+            game.reset_game()
+            print("reseting game!")
```
Comment (frankbryden) [inline backend/main.py]: This doesn't take into account the wait time between games

File: backend/main.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                await websocket.send_text(f"A game is not currently in progress.")
+            else:
+                amount = message["amount"]
+                await manager.broadcast(f"{current_players[ws]} bid {amount}.")
+        
+        elif message["type"] == "cashout":
+            if not game.ongoing:
+                await websocket.send_text(f"A game is not currently in progress.")
+            else:
+                await manager.broadcast(f"{current_players[ws]} cashed out with a mult of {game.get_multiplicator()}!")
```
Comment (frankbryden) [inline backend/main.py]: These should all be sending JSON please 🙏

Comment (frankbryden) [reply backend/main.py]: (following the format agreed upon in that comment)

File: backend/main.py (hunk 1 above)
Comment (Choklatier) [reply backend/main.py:45]: Used the threading module to get event-based handling for when there is no player.

File: backend/main.py [hunk 4]
```diff
@@ -21,33 +22,59 @@ async def connect(self, websocket: WebSocket):
...
-    async def send_personal_message(self, message: str, websocket: WebSocket):
-        await websocket.send_text(message)
+    async def send_personal_message_lobby(self, message: str, websocket: WebSocket):
+        await websocket.send_text(str({"type":"lobby","message":message}))
 
-    async def broadcast(self, message: str):
+    async def broadcast_lobby(self, message: str):
         for connection in self.active_connections:
-            await connection.send_text(message)
+            await connection.send_text(str({"type":"lobby","message":message}))
```
Comment (frankbryden) [inline backend/main.py]: This should be a call to `json.dumps()`, not `str()`

Comment (frankbryden) [reply backend/main.py]:
```suggestion
            await connection.send_text(json.dumps({"type":"lobby","message":message}))
```

File: backend/main.py [hunk 5]
```diff
@@ -21,33 +22,59 @@ async def connect(self, websocket: WebSocket):
     def disconnect(self, websocket: WebSocket):
         self.active_connections.remove(websocket)
 
-    async def send_personal_message(self, message: str, websocket: WebSocket):
-        await websocket.send_text(message)
+    async def send_personal_message_lobby(self, message: str, websocket: WebSocket):
+        await websocket.send_text(str({"type":"lobby","message":message}))
```
Comment (frankbryden) [inline backend/main.py]:
```suggestion
        await websocket.send_text(json.dumps({"type":"lobby","message":message}))
```

File: backend/main.py [hunk 6]
```diff
@@ -21,33 +22,59 @@ async def connect(self, websocket: WebSocket):
...
 
 
 manager = ConnectionManager()
 
 game_handler = GameHandler()
 game = game_handler.get_game()
 
+# Event from threading library to get event driven approach :D
+# The event is set in the websocket handler for when a player joins
+player_join_event = Event()
```
Comment (frankbryden) [inline backend/main.py]: ohhh amazing!! Great !

File: backend/main.py [hunk 7]
```diff
@@ -21,33 +22,59 @@ async def connect(self, websocket: WebSocket):
...
+async def loop():
     while True:
         current_players = game_handler.players
-        
-        if len(current_players) > 0 and not game.ongoing:
+
+        # If there is no player we wait for one to save CPU cycles
+        if len(current_players) == 0:
+            print("Waiting for players...")
+            player_join_event.wait()
```
Comment (frankbryden) [inline backend/main.py:51]: this is awesome! For real!

File: backend/main.py [hunk 8]
```diff
@@ -63,31 +90,34 @@ async def websocket_endpoint(websocket: WebSocket):
...
+                    await manager.send_personal_message_lobby(f"A game is not currently in progress.")
+                else:
+                    amount = message["amount"]
+                    await manager.broadcast_lobby(f"{current_players[websocket]} bid {amount}.")
+            
+            elif message["type"] == "cashout":
+                if not game.ongoing:
+                    await manager.send_personal_message_lobby(f"A game is not currently in progress.")
+                else:
+                    await manager.broadcast_lobby(f"{current_players[websocket]} cashed out with a mult of {game.get_multiplicator()}!")
```
Comment (frankbryden) [inline backend/main.py]: Sorry Theo for complaining again about this.
But the messages need to be "structured", in the sense they need to be understood by the frontend so the events can be rendered.

For example, a join event might look something like (just a suggestion):
```
{
  "type": "join",
  "target": "Theo",
  "lobby": ["Nicolas", "Theo"]
}
```
(btw on a sidenote - I realize the `join` event is one that goes both client -> server and server -> client and has different meanings. On the `client -> server`, it's a "I'm here to join the lobby". On the `server -> client`, it's a "hey guys, this new person has just joined".

This message is parsed by the frontend, which then updates the lobby to render the updated state, perhaps printing a message saying `Theo joined`

File: backend/game_handler.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
+import numpy as np
+import time
+from threading import Lock
+
+class Game:
+    def __init__(self,name : str,starting_multiplicator : float = 0.75,multiplicator_coef : float = 0.01,
+                 average : float = 30,standard_deviation : float = 10, minimum_time : float = 3,
+                 players : list = None):
```
Comment (frankbryden) [inline backend/game_handler.py]:
```suggestion
    def __init__(self,name: str, starting_multiplicator: float = 0.75, multiplicator_coef: float = 0.01,
                 average: float = 30, standard_deviation : float = 10, minimum_time: float = 3,
                 players: list = None):
```

File: backend/game_handler.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    def is_crashed(self) -> bool:
+        if self.ongoing:
+            return (time.time() - self.initial_time) > self.game_duration
+        else:
+            return False # Game is not crashed, it has not started
+
+class GameHandler:
+    
+    def __init__(self, name : str = "New Lobby", average : float = 30,standard_deviation : float = 10,
+                 minimum_time : float = 3, starting_multiplicator : float = 0.75,multiplicator_coef : float = 0.01):
```
Comment (frankbryden) [inline backend/game_handler.py]:
```suggestion
    def __init__(self, name: str = "New Lobby", average: float = 30, standard_deviation: float = 10,
                 minimum_time: float = 3, starting_multiplicator: float = 0.75, multiplicator_coef: float = 0.01):
```

File: backend/game_handler.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        self.mutex = Lock()
+    
+    # only for some testing
+    def test_values(self) -> tuple: 
+        duration = max(self.min_time,np.random.normal(self.average,self.std))
+        time_array = np.linspace(0,duration)
+        multiplicator_array = self.starting_multiplicator + np.exp( self.multiplicator_coef * time_array)
+        return duration,multiplicator_array
+
+    def get_game(self,name : str = "New Game")-> Game:
```
Comment (frankbryden) [inline backend/game_handler.py]:
```suggestion
    def get_game(self, name: str = "New Game")-> Game:
```

File: backend/game_handler.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+         
+        return Game(name,
+                    starting_multiplicator = self.starting_multiplicator,
+                    multiplicator_coef = self.multiplicator_coef,
+                    average = self.average,standard_deviation = self.std,
```
Comment (frankbryden) [inline backend/game_handler.py]:
```suggestion
                    average = self.average, standard_deviation = self.std,
```

File: backend/game_handler.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+            self.players[ws] = name
+
+
+if __name__ == "__main__":
+    import matplotlib.pyplot as plt
+    game_handler = GameHandler()
+    durations = []
+    multiplicators = []
+    for i in range(10000):
+        duration,multiplicator_array = game_handler.test_values()
```
Comment (frankbryden) [inline backend/game_handler.py]:
```suggestion
        duration, multiplicator_array = game_handler.test_values()
```

**Review by frankbryden — APPROVED** [review]
LGTM

**Pull request merged**